### PR TITLE
Feature/colocate

### LINKED ISSUE #315: Add colocation method in Field
## Description
The figure below depicts what grid we decided to use for the Model, State and Increment.
SOCA-Horizontal-Grids

Adding the increment to the state requires colocating the increment to the state's grid.

## Solution
Add a colocation method to the field object.

## Related to
Partially addresses #282

### PR DESCRIPTION
**Pull request by guillaumevernieres**
## Description
Colocate increment to state in `add_incr`. Interpolation uses fms idw ... hu-hoooo .... For now.
**bug fix:** Geostrophic increment for u-v was added without colocation.
Change of answer expected in the 3dvar_godas test, which makes use of geostrophy.

### Issue(s) addressed
closes #315

## Testing
How were these changes tested? **Colocation called inside of 3dvar_godas**
What compilers / HPCs was it tested with? **gcc**
Are the changes covered by ctests? **yes**

## Dependencies
- ~[oops PR#585](https://github.com/JCSDA/oops/pull/585)~
Deferred to an other PR, we'll just have to remove the NOCOMPARE in the test and add the reference file.

### CODE REVIEW

**Comment by guillaumevernieres** [comment]
I'll blame the multiple test failure on the coffee rationing. I know the reviewers are eager to start plunging into the added code (it's quite amazing if I may say), so I'll be pushing the fix in a few minutes.

File: src/soca/Fields/soca_fields_mod.F90 [hunk 1]
```diff
@@ -1060,4 +1099,73 @@ subroutine soca_fields_write_rst(fld, f_conf, vdate)
...
+
+  ! Apply interpolation to all fields, when necessary
+  do i=1,size(self%fields)
+
+    ! Check if already colocated
+    if (self%fields(i)%c_grid_loc == cgridlocout) cycle
+
+    ! Initialize fms spherical idw interpolation
+     g => self%geom
+     call horiz_interp_spherical_new(interp2d, &
```
Comment (travissluka) [inline src/soca/Fields/soca_fields_mod.F90:1140]: memory leak, you need a matching call to `horiz_interp_spherical_del` at least once at the end
(If you look inside `horiz_interp_type` all the data members are pointers and so wont automatically be deallocated upon `interp2d` deallocation)

**Review by travissluka — CHANGES_REQUESTED** [review]

File: src/soca/Fields/soca_fields_mod.F90 (hunk 1 above)
Comment (guillaumevernieres) [reply src/soca/Fields/soca_fields_mod.F90:1140]: Arf, yes, good catch, forgot the cleaning. I'll spend "some" time on the coverage too.

Comment (travissluka) [reply src/soca/Fields/soca_fields_mod.F90:1140]: :+1: is there another way we we could test it? I'm feel uneasy using `3dvar_godas` as the dumping ground for testing new features

Comment (guillaumevernieres) [reply src/soca/Fields/soca_fields_mod.F90:1140]: > +1 is there another way we we could test it? I'm feel uneasy using `3dvar_godas` as the dumping ground for testing new features

Ideally it should be tested in the oops State test through the testing of `add_incr`, but `add_incr` is not tested. Maybe I'll add one tomorrow while I procrastinate over my "smart" goals. This doesn't resolve the issue of poor coverage (the test only exercising the colocation to the h-grid). I'll think about it tomorrow, but maybe I need to remove the u-grid and v-grid options.

**Comment by guillaumevernieres** [comment]
The add incr test will fail until [oops PR#585](https://github.com/JCSDA/oops/pull/585) is merged, or someone that's not @guillaumevernieres tells travis-ci to build with the bugfix branch in oops ...

**Comment by guillaumevernieres** [comment]
- Commented out the u and v colocation option until we figure out a way to test these options
- Removed the dependency to [oops pr585](https://github.com/JCSDA/oops/pull/585) by removing the compare test in the `addincrement` test (and realizing as I am typing that I forgot to remove the testref file ...)
- @travissluka I'm OK using 3dvar_godas to test new features, why the uneasiness?

**Comment by travissluka** [comment]
@guillaumevernieres It's called a "unit test" for a reason. Throwing everything in 3dvar_godas no longer makes it a "unit". If the answers change, we have a harder time tracking down why given that there are multiple features that aren't tested elsewhere. But whatever, deal with that later

**Comment by travissluka** [comment]
... you never committed your addincrement source files.

**Comment by guillaumevernieres** [comment]
@travissluka 
- Unit test for rotate is in varchange and convert state
- Unit test for colocate is in addincrement
3dvar_godas makes use of these new features, the same way is uses a bunch of other stuff. But you are correct, it was not tested before my unfortunate latest push that was suppose to have the `addincrement` test ... My bad, will push in a few minutes.

**Comment by travissluka** [comment]
:+1: yeah my comment about testing was before `addincrement`

**Review by travissluka — APPROVED** [review]

**Pull request merged**